native/dgeqp3.go

```diff
@@ -0,0 +1,170 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+// Dgeqp3 computes a QR factorization with column pivoting of the
+// m×n matrix A: A*P = Q*R using Level 3 BLAS.
+//
+// The matrix Q is represented as a product of elementary reflectors
+//  Q = H_0 H_1 . . . H_k-1, where k = min(m,n).
+// Each H[i] has the form
+//  H[i] = I - tau * v * v^T
+// where tau is a real vector, and v is a real/complex vector
+// with v[0:i-1] = 0 and v[i] = 1; v[i:m] is stored on exit in
+// A[i:m, i], and tau in tau[i].
+//
+// jpvt specifies a column pivot to be applied to A. If
+// jpvt[j] is at least zero, the jth column of A is permuted
+// to the front of A*P (a leading column), otherwise the jth
+// column of A is a free column. On return, jpvt holds the
+// permutation that was applied; if jpvt[j] == k, then the jth
+// column of A*P was the the kth column of A.
+// jpvt must have length n or Dgeqp3 will panic.
+//
+// On exit tau holds the scalar factors of the elementary reflectors.
+// It must have length n, otherwise Dgeqp3 will panic.
+//
+// work must have length max(1, lwork). otherwise Dgeqp3 will panic.
+// work must have length at least max(1,lwork), and lwork must be at least
+// 3*n+1. For optimal performance lwork must be at least 2*n+(n+1)*nb, where
+// nb is the optimal blocksize.On return, work[0] will contain the optimal
+// value of lwork.
+//
+// If lwork == -1, instead of performing Dgeqp3, only the optimal value of lwork
+// will be stored in work[0].
+//
+// Dgeqp3 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dgeqp3(m, n int, a []float64, lda int, jpvt []int, tau, work []float64, lwork int) {
+	const (
+		inb    = 1
+		inbmin = 2
+		ixover = 3
+	)
+	checkMatrix(m, n, a, lda)
+
+	if len(jpvt) != n {
+		panic(badIpiv)
+	}
+	minmn := min(m, n)
+	if len(tau) < minmn {
+		panic(badTau)
+	}
+	if len(work) < max(1, lwork) {
+		panic(badWork)
+	}
+
+	var iws, lwkopt, nb int
+	if minmn == 0 {
+		iws = 1
+		lwkopt = 1
+	} else {
+		iws = 3*n + 1
+		nb = impl.Ilaenv(inb, "DGEQRF", " ", m, n, -1, -1)
+		lwkopt = 2*n + (n+1)*nb
+	}
+	work[0] = float64(lwkopt)
+
+	if lwork == -1 {
+		return
+	}
+
+	bi := blas64.Implementation()
+
+	// Move initial columns up front.
+	nxfd := 0
+	for j := 0; j < n; j++ {
+		if jpvt[j] < 0 {
+			jpvt[j] = j
+			continue
+		}
+		if j != nxfd {
+			bi.Dswap(m, a[j:], lda, a[nxfd:], lda)
+			jpvt[j], jpvt[nxfd] = jpvt[nxfd], j
+		} else {
+			jpvt[j] = j
+		}
+		nxfd++
+	}
+
+	// Factorize fixed columns.
+	//
+	// Compute the QR factorization of fixed columns and update remaining columns.
+	if nxfd >= 0 {
+		na := min(m, nxfd)
+		impl.Dgeqrf(m, na, a, lda, tau, work, lwork)
+		iws = max(iws, int(work[0]))
```

Comment:
Our native implementation does not set work[0] to optimum lwork. That should be fixed. Issue filed.